File: typescript/src/integrations/_shared/filters.ts
/**
 * Name-based eligibility filter for tool / middleware integrations.
 *
 * Mirrors Python `_shared/filters.py`.
 */

export interface FilterOptions {
  allow?: Iterable<string>;
  ignore?: Iterable<string>;
}

/**
 * Build a name predicate from allow- or ignore-lists.
 *
 * - `allow`:  only names in this set return true.
 * - `ignore`: every name except those in this set returns true.
 * - both undefined: every name returns true (no-op).
 *
 * Passing both throws — pick one mode.
 */
export function makeFilter(
  options: FilterOptions = {}
): (name?: string | null) => boolean {
  const { allow, ignore } = options;
  if (allow !== undefined && ignore !== undefined) {
    throw new Error('Pass `allow` OR `ignore`, not both.');
  }
  if (allow !== undefined) {
    const allowed = new Set(allow);
    return (name) => name != null && allowed.has(name);
  }
  if (ignore !== undefined) {
    const ignored = new Set(ignore);
    return (name) => name == null || !ignored.has(name);
  }
  return () => true;
}

File: python/compresr/integrations/_shared/filters.py
"""Name-based eligibility filter used by tool/middleware integrations."""

from __future__ import annotations

from typing import Callable, Iterable, Optional


def make_filter(
    allow: Optional[Iterable[str]] = None,
    ignore: Optional[Iterable[str]] = None,
) -> Callable[[Optional[str]], bool]:
    """Build a name predicate from allow- or ignore-lists.

    - ``allow``: only names in this set return True.
    - ``ignore``: every name except those in this set returns True.
    - Both ``None``: every name returns True (no-op).

    Passing both raises ``ValueError`` — pick one mode.
    """
    if allow is not None and ignore is not None:
        raise ValueError("Pass allow OR ignore, not both.")

    if allow is not None:
        allowed = frozenset(allow)
        return lambda name: name is not None and name in allowed

    if ignore is not None:
        ignored = frozenset(ignore)
        return lambda name: name is None or name not in ignored

    return lambda _name: True
